### PR TITLE
feat: #4160 #4162 deployment service

### PR DESCRIPTION
**Pull request by bashleigh**
# Pull request checklist

**Detail as per issue below (required):**

fixes: #4160 
fixes: #4162

### CODE REVIEW

File: packages/api-key-service/package.json [hunk 1]
```diff
@@ -0,0 +1,14 @@
+{
+  "name": "api-key",
+  "dependencies": {
+    "@aws/dynamodb-data-mapper-annotations": "^0.7.3",
+    "@homeservenow/serverless-aws-handler": "^0.2.4",
+    "class-transformer": "^0.4.0",
+    "class-validator": "^0.13.1"
+  },
+  "scripts": {
```
Comment (willmcvay) [inline packages/api-key-service/package.json:9]: Will expect a build, release and publish script too - you can echo skip them obvs, just makes the pipeline simple ie "run these tasks in order"

File: packages/deployment-service/src/utils/decode.token.ts [hunk 2]
```diff
@@ -0,0 +1,10 @@
+import jwt from 'jsonwebtoken'
+
+export const decodeToken = (
```
Comment (willmcvay) [inline packages/deployment-service/src/utils/decode.token.ts:3]: We already have one of these - it also validates the token! See above

File: packages/api-key-service/src/functions/create.api-key.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+import { ApiKeyDto } from '@/dto'
+import { ApiKeyModel } from '@/models'
+import { createApiKey as create } from '@/services'
+import { authorised, decodeToken } from '@reapit/node-utils'
+
+export const createApiKey = httpHandler<ApiKeyDto, ApiKeyModel>({
+  serialise: {
+    input: (event): ApiKeyDto => {
+      authorised(event)
+      const customer = decodeToken(event.headers['reapit-connect-token'] as string)
```
Comment (willmcvay) [inline packages/api-key-service/src/functions/create.api-key.ts:13]: For decoding tokens you can use this https://github.com/reapit/foundations/blob/master/packages/connect-session/src/utils/verify-decode-id-token.ts#L79

Usage `import {connectSessionVerifyDecodeIdToken } from '@reapit/connect-session'`

Comment (willmcvay) [reply packages/api-key-service/src/functions/create.api-key.ts:13]: Returns this interface https://github.com/reapit/foundations/blob/master/packages/connect-session/src/types.ts#L22

File: packages/cli/package.json [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  "devDependencies": {
+    "@types/figlet": "^1.5.1",
+    "@types/inquirer": "^7.3.1",
+    "@types/yargs": "^16.0.2",
+    "concurrently": "^6.1.0",
+    "eslint": "^7.26.0",
+    "prettier": "^2.3.0",
+    "ts-node": "^9.1.1"
+  },
+  "scripts": {
```
Comment (willmcvay) [inline packages/cli/package.json:23]: As above, pipeline will expect all commands to be implemented even if skipped

File: packages/cli/README.md [hunk 5]
```diff
@@ -0,0 +1 @@
+# cli
```
Comment (willmcvay) [inline packages/cli/README.md:1]: Can we call the package something a bit more helpful eg `deployment-cli`? 😁

File: packages/deployment-service/src/functions/create.deployment.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
+import { httpHandler, ValidationException } from '@homeservenow/serverless-aws-handler'
+import { DeploymentDto } from '@/dto'
+import { DeploymentModel } from '@/models'
+import * as service from '@/services/deployment'
+import { plainToClass, classToClassFromExist } from 'class-transformer'
+import { validate } from 'class-validator'
+import { authorised } from '@/utils'
+import { decodeToken } from '@/utils/decode.token'
```
Comment (willmcvay) [inline packages/deployment-service/src/functions/create.deployment.ts:8]: See above decoding tokens

File: packages/deployment-service/src/functions/create.pipeline.spec.ts [hunk 7]
```diff
@@ -0,0 +1,21 @@
+import { createPipeline } from './create.pipeline'
```
Comment (willmcvay) [inline packages/deployment-service/src/functions/create.pipeline.spec.ts:1]: It's a small point but can tests live in a `__tests__` folder please? It's just the convention in the repo

File: packages/deployment-service/src/functions/paginate.deployment.ts [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+ * Return pagination response for signed in user
+ */
+export const paginateDeployments = httpHandler({
+  serialise: {
+    input: (event) => {
+      authorised(event)
+    },
+  },
+  handler: async ({ event }): Promise<QueryPaginator<DeploymentModel>> => {
+    const customer = decodeToken(event.headers['reapit-connect-token'] as string)
```
Comment (willmcvay) [inline packages/deployment-service/src/functions/paginate.deployment.ts:17]: Decode token etc

File: tsconfig.json [hunk 9]
```diff
@@ -4,8 +4,8 @@
     "esModuleInterop": true,
     "allowSyntheticDefaultImports": true,
     "jsx": "react",
-    "lib": ["dom", "es2015", "es2017", "es2018", "es2019", "es2020", "es5", "es6", "es7", "esnext"],
-    "module": "es2020",
+    "lib": ["dom", "es2015", "es2017", "es2018", "es2019", "es5", "es6", "es7", "esnext"],
```
Comment (willmcvay) [inline tsconfig.json:7]: Did you mean to remove this? Should tidy up the comment if so - be wary, may have effects on other packages!

File: packages/utils/src/node/authorisation/decode.token.ts [hunk 10]
```diff
@@ -0,0 +1,10 @@
+import jwt from 'jsonwebtoken'
+
+export const decodeToken = (
```
Comment (willmcvay) [inline packages/utils/src/node/authorisation/decode.token.ts:3]: Like I say, I don't think we need this func - we should validate tokens too and the existing one does this

**Review by willmcvay — CHANGES_REQUESTED** [review]
So just a general comment - quite light on tests - suggest we add to the ticket list to review later -  don't want to block this getting merged

File: packages/deployment-service/src/utils/decode.token.ts (hunk 2 above)
Comment (bashleigh) [reply packages/deployment-service/src/utils/decode.token.ts]: deleted

File: packages/api-key-service/src/functions/create.api-key.ts (hunk 3 above)
Comment (bashleigh) [reply packages/api-key-service/src/functions/create.api-key.ts]: deleted

File: packages/cli/README.md (hunk 5 above)
Comment (bashleigh) [reply packages/cli/README.md:1]: was thinking reapit-cli

File: packages/deployment-service/src/functions/create.deployment.ts (hunk 6 above)
Comment (bashleigh) [reply packages/deployment-service/src/functions/create.deployment.ts]: deleted

File: packages/api-key-service/package.json [hunk 11]
```diff
@@ -0,0 +1,25 @@
+{
+  "name": "api-key",
```
Comment (willmcvay) [inline packages/api-key-service/package.json:2]: Should be api-key-service - yarn uses the name for workspaces

File: packages/cli/README.md [hunk 12]
```diff
@@ -0,0 +1,4 @@
+# reapit
```
Comment (willmcvay) [inline packages/cli/README.md:1]: You were going to change the package name from cli to developer-cli

File: packages/api-key-service/package.json (hunk 11 above)
Comment (willmcvay) [inline packages/api-key-service/package.json:2]: Think this needs to be api-key-service to match the directory - it's to do with how yarn workspaces work

**Review by willmcvay — CHANGES_REQUESTED** [review]
Just a couple of small things...

File: packages/api-key-service/package.json (hunk 11 above)
Comment (willmcvay) [reply packages/api-key-service/package.json:2]: Don't know why but this comment is duplicated!

Comment (bashleigh) [reply packages/api-key-service/package.json:2]: sure, I'll change it, did think this when I was watching the tests

**Review by willmcvay — APPROVED** [review]
💯

**Comment by bashleigh** [comment]
100 commits exact too 😂

**Pull request merged**